### PR TITLE
MAINT: When linspace's step is a NumPy scalar, do multiplication in-place

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
This is faster and more memory efficient.

Out-of-place operation was preventing me from using linspace
to create over 2**31 equispaced doubles on my machine due to
insufficient memory.

Is this sufficient to no re-break [gh-7142](https://github.com/numpy/numpy/issues/7142)?

### CODE REVIEW

**Comment by eric-wieser** [comment]
Can you explain why you believe this to be more memory efficient? None of the objects which pass `isscalar` implement `__imul__` anyway:
```
>>> [x for x in np.ScalarType if hasattr(x, '__imul__')]
[<class 'numpy.object_'>]
```

**Comment by eric-wieser** [comment]
Nevermind, I'm reading that backwards

File: numpy/core/function_base.py [hunk 1]
```diff
@@ -125,7 +125,10 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None):
             # One might be tempted to use faster, in-place multiplication here,
             # but this prevents step from overriding what class is produced,
             # and thus prevents, e.g., use of Quantities; see gh-7142.
-            y = y * step
+            if np.isscalar(step):
```
Comment (eric-wieser) [inline numpy/core/function_base.py]: `np` isn't imported here, which makes all the tests fail

**Comment by eric-wieser** [comment]
This seems reasonable, and at a glance doesn't appear to break reasonable subclasses. You'll need to update the surrounding comment though.

**Comment by mhvk** [comment]
I tried to think whether this could break code again, but think even a custom subclass of `generic` for `start` or `stop` would be OK. I would explicitly test `if isinstance(step, np.generic)` here, since given that things went through `asanyarray`, that is the only thing a regular number could have turned into.

Separately, perhaps you want to do the test on `delta` rather than `step`, so that both branches can be optimized.

**Comment by oleksandr-pavlyk** [comment]
@mhvk I modified the patch to test delta, and multiply in place when ok to do so in both branch within the `num > 1` branch. The `num == 1` complementary branch works with length 1 vector `y` where out-of-place multiplication is not a cause for excessive memory footprint.

File: numpy/core/function_base.py [hunk 2]
```diff
@@ -115,17 +115,24 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None):
...
+                y = y * delta
         else:
-            # One might be tempted to use faster, in-place multiplication here,
-            # but this prevents step from overriding what class is produced,
-            # and thus prevents, e.g., use of Quantities; see gh-7142.
-            y = y * step
+            if _mult_inplace:
+                y *= step
+            else:
+                # One might be tempted to use faster, in-place multiplication here,
```
Comment (mhvk) [inline numpy/core/function_base.py]: Maybe move the comment up to where `_mult_inplace` is defined and rephrase it as "In-place multiplication is faster, but  prevents the multiplicant from overriding what class is produced. Hence, we do it only for standard scalar types."

**Review by mhvk — COMMENTED** [review]
This looks good; just one comment about the placement of the rationale for not doing in-place always. Ideally, there would be a test, but I don't really see how to do that, and the code is quite obvious.

While you implement that, could you squash or rebase to get to one commit. I guess it is not strictly a bug, so the first word should be `MAINT`

File: numpy/core/function_base.py [hunk 3]
```diff
@@ -115,17 +115,24 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None):
     y = _nx.arange(0, num, dtype=dt)
 
     delta = stop - start
+    # In-place multiplication y *= delta/div is faster, but prevents the multiplicant
+    # from overriding what class is produced, and thus prevents, e.g. use of Quantities,sSee gh-7142,
```
Comment (mhvk) [inline numpy/core/function_base.py]: Sorry to ask for yet another iteration, but the copy/pasting seems to have gone wrong here... Just move "See gh-7142" to the next line...

Comment (oleksandr-pavlyk) [reply numpy/core/function_base.py]: @mhvk Not a problem, sorry for the sloppiness. I have forced pushed an amendment.

**Comment by mhvk** [comment]
OK, thanks! Merging...

**Pull request merged**